File: code/F-SER/traffic.f90
program traffic

  use trafficlib

  implicit none

! Set the size of the road

  integer :: ncell = 10240000
  integer :: maxiter, printfreq

  integer :: i, iter, nmove, ncars
  real    :: density

  integer, allocatable, dimension(:) :: newroad, oldroad

  double precision :: tstart, tstop

  maxiter = 1.024e9/float(ncell)
  printfreq = maxiter/10

! Set target density of cars

  density = 0.52

  write(*,*) 'Length of road is ', ncell
  write(*,*) 'Number of iterations is ', maxiter
  write(*,*) 'Target density of cars is ', density

! Allocate arrays

  allocate(newroad(0:ncell+1))
  allocate(oldroad(0:ncell+1))

! Initialise road accordingly using random number generator

  do i = 1, ncell
     oldroad(i) = 0
     newroad(i) = 0
  end do

  write(*,*) 'Initialising ...'

  ncars = initroad(oldroad(1), ncell, density, seed)

  write(*,*) '... done'

  write(*,*) 'Actual density of cars is ', float(ncars)/float(ncell)
  write(*,*)

  tstart = gettime()

  do iter = 1, maxiter

     call updatebcs(oldroad, ncell)

     nmove = updateroad(newroad, oldroad, ncell)

! Copy new to old array

     do i = 1, ncell

        oldroad(i) = newroad(i)

     end do

     if (mod(iter, printfreq) .eq. 0) then

        write(*,*) 'At iteration ', iter, ' average velocity is ', &
             float(nmove)/float(ncars)
     end if

  end do

  tstop = gettime()

  deallocate(newroad)
  deallocate(oldroad)

  write(*,*)
  write(*,*) 'Finished'
  write(*,*)
  write(*,*) 'Time taken was  ', tstop - tstart, ' seconds'
  write(*,*) 'Update rate was ', &
              1.d-6*float(ncell)*float(maxiter)/(tstop-tstart), ' MCOPs'
  write(*,*)

end program traffic

File: notebooks/traffic/traffic.py
#!/usr/bin/env python

import sys
import time

from trafficlib import initroad, updatebcs, updateroad, gettime

def main(argv):

    # Simulation parameters
    seedval = 5743
    ncell = 10240000
    maxiter = 1024000000//ncell
    printfreq = maxiter//10

    tmproad = [0 for col in range(ncell)]
    newroad = [0 for col in range(ncell+2)]
    oldroad = [0 for col in range(ncell+2)]

    density = 0.52

    print(f"Length of road is {ncell}")
    print(f"Number of iterations is {maxiter}")
    print(f"Target density of cars is {density}")

    # Initialise road accordingly using random number generator
    print(f"Initialising ...")

    ncars = initroad(tmproad, density, seedval)

    print(f"Actual Density of cars is {float(ncars)/float(ncell)}\n")

    oldroad[1:ncell] = tmproad
    
    tstart = gettime()

    for iter in range(1, maxiter+1):

        updatebcs(oldroad)

        nmove = updateroad(newroad, oldroad)

        # Copy new to old array
        for i in range(1, ncell+1):
            oldroad[i] = newroad[i]

        if iter % printfreq == 0:

          print(f"At iteration {iter} average velocity is {float(nmove)/float(ncars):.6f}")

    tstop = gettime()

    print(f"\nFinished\n")
    print(f"Time taken was {tstop-tstart:.2f} seconds")
    print(f"Update rate was {1.0e-6*ncell*maxiter/(tstop-tstart):.2f} MCOPs")

if __name__ == "__main__":
    main(sys.argv[1:])
